### DISCUSSION TITLE
How to adjust structure folder for monorepositories?

### DISCUSSION THREAD

**Discussion by flo-sch** [Azure/azure-functions-nodejs-worker#481] [General]
I work with typescript and [nx.dev](https://nx.dev/) workspace: a build tool allowing to easily lint, test, build and share source code inside a monorepo (using TS path mapping from a root `tsconfig.base.json`).

It works with a specific folder structure which makes it a bit hard to combine with Azure Functions:

    /tsconfig.base.json
    /package.json
    /node_modules
    /host.json (?)
    /.funcignore (?)
    /apps
      /app1
      /app2
      /app3
    /libs
      /lib1
      /lib2
    /tools
      ...

As far as I know, Azure Functions Core Tools only parse one level of subfolders to find `function.json` files (for `func start` or `func azure functionapp publish <name>`).
Unfortunately, since nx works with an `apps` main subfolder, I cannot put the `function.json` inside each "app" and have to create a bunch subfolders in the root level to put the `function.json` for each app, adding a `scriptFile` to reference the output of `nx serve`.

Is there any way to configure the core tools to look for `function.json` inside `./apps/*/function.json` instead?

I have read in the issues here about `--prefix` and `--script-root` flags but I did not manage to make it work (and do not find those documented neither)

Thanks a bunch in advance 🙏 

Example repository here: https://github.com/flo-sch/nx-workspace-azure-function

Issue related to:
- https://github.com/Azure/azure-functions-core-tools/issues/858
- https://github.com/Azure/Azure-Functions/issues/1240

**Comment by ejizba** [comment]
Hi @flo-sch unfortunately there's no good answer for this at the moment. However, we're actively working on a solution for this. Some changes were recently made on the host here: https://github.com/Azure/azure-functions-host/pull/7606 and I've created this issue to track the Node.js changes here https://github.com/Azure/azure-functions-nodejs-worker/issues/480

We definitely plan to address allowing "function.json" in other folder structures. However, I've converted this "issue" into a "discussion" since we're still in the design phase and I was wondering if you had any other feedback. For example, we noticed each "apps/function1" folder has it's own tsconfig.json - can you give a little reasoning why that's necessary?

**Reply by flo-sch** [reply]
Disclaimer: such a structure comes from `nx` itself, I did not come with it myself, I can ask in their Slack if someone knows better, but my 20cents feeling is towards optimisation: scaling apps & libs at scale can quickly become greedy for ts especially in watch mode, and AFAIK those configs here purely turn rules on and off for some specific patterns.

(Though it can be achieved by a single tsconfig file as well... Flexibility perhaps?)

**Comment by flo-sch** [comment]
Hello @ejizba, coming back here after quite some time, I forgot a bit about this thread.

I tested the framework v4 ([pull-request available there](https://github.com/flo-sch/nx-workspace-azure-function/pull/1)) and it seems to work fine:
A few findings, food for thoughts:

1. Removing the subfolder containing functions.json files simplifies the folder structure (potentially a lot!) ✅ 

2. Using `"main": "dist/apps/*/main.js"` works for me at the moment, I just hope it will not conflict with other tools in the future (as `package.json` in a monorepository may not only serve the Function App, potentially frontend applications and/or libraries etc.)

3. The possibility to create a test InvocationContext is great ✅

4. The features coming from https://github.com/Azure/azure-functions-nodejs-library/pull/46 are great (`jsonBody` in particular for me) ✅ 

I am looking forward a stable release of that v4 🚀
